eip4844 beacon block proposals

henridf opened the pull request:
Still need to add nim-kzg library and use it to compute the aggregated proof.

henridf committed:
eip4844 beacon block proposals

tersec commented on beacon_chain/validators/validator_duties.nim:
```diff
@@ -437,6 +440,32 @@ proc getExecutionPayload[T](
+        default(BlobsBundleV1)
+  except CatchableError as err:
+    beacon_block_payload_errors.inc()
+    warn "Getting execution payload from Engine API failed",
```
Mismatched warning message? This is about blobs, not execution payloads

tersec commented on beacon_chain/validators/validator_duties.nim:
```diff
@@ -437,6 +440,32 @@ proc getExecutionPayload[T](
+proc getBlobsBundle(
+    node: BeaconNode, epoch: Epoch, validator_index: ValidatorIndex,
+    payload_id: PayloadID): Future[BlobsBundleV1] {.async.} =
+  # https://github.com/ethereum/consensus-specs/blob/v1.1.10/specs/bellatrix/validator.md#executionpayload
```
Not relevant spec URL?

henridf replied:
oops, fixed

tersec commented on beacon_chain/spec/state_transition.nim:
```diff
@@ -482,10 +486,11 @@ template partialBeaconBlock*(
       ))
 
-proc makeBeaconBlock*[T: bellatrix.ExecutionPayload | capella.ExecutionPayload](
+proc makeBeaconBlock*[T: bellatrix.ExecutionPayload | capella.ExecutionPayload | eip4844.ExecutionPayload](
```
Long line

henridf replied:
fixed

tersec commented:
failing on `make wss_sim` -- needs the KZG commitment parameters added there too

tersec commented:
failing on `make wss_sim` -- needs the KZG commitment parameters added there too

henridf commented:
> needs the KZG commitment parameters added there too
done

henridf commented:
> Still need to add nim-kzg library and use it to compute the aggregated proof.

My preference would be to do that in a follow-on PR (presumably tomorrow), fwiw. I think in the current state, we'd propose invalid eip4844 blocks, but no behavior should change for pre-eip4844.

tersec commented:
> > Still need to add nim-kzg library and use it to compute the aggregated proof.
> 
> My preference would be to do that in a follow-on PR (presumably tomorrow), fwiw. I think in the current state, we'd propose invalid eip4844 blocks, but no behavior should change for pre-eip4844.

My concern is that I suspect this won't pass Jenkins testnet finalization at all because of the `minimal` preset issues with EIP4844 you discovered. It hasn't gotten that far yet, and I'm not really sure why. I'm fine with merging this as long as it does something sensible with EIP4844 minimal presets and engine API calls.

(To summarize: `BlobsBundleV1` in https://github.com/ethereum/execution-apis/blob/main/src/engine/experimental/blob-extension.md#blobsbundlev1  assumes a `mainnet` preset blob size, where https://github.com/ethereum/consensus-specs/blob/dev/presets/minimal/eip4844.yaml defines `FIELD_ELEMENTS_PER_BLOB: 4`, which leads to a Blob according to https://github.com/ethereum/consensus-specs/blob/dev/specs/eip4844/polynomial-commitments.md#custom-types having a length of `4 * 32`, not `4096 * 32` which `BlobsBundleV1` requires.)

tersec commented:
I agree, `nim-kzg` should be in another PR though, not this one.

henridf committed:
Don't fetch blobs under minimal preset

@tersec's summary of the issue:

BlobsBundleV1 in the execution API spec assumes a mainnet preset blob
size, where the EIP4844 consensus spec defines
FIELD_ELEMENTS_PER_BLOB: 4 under the minimal preset, which leads to a
Blob having a length of 4 * 32, not 4096 * 32 which BlobsBundleV1
requires.

henridf commented:
> something sensible with EIP4844 minimal presets and engine API calls.

Following your suggestion, made a change 2420d69011879610ae012b4a078091bc7fe99c5f to skip blob fetching when the preset is minimal. This means that all block proposals will be invalid for EIP4844-minimal, but as we've agreed elsewhere, EIP4844 is mainnet-preset only for now.

tersec commented:
> > something sensible with EIP4844 minimal presets and engine API calls.
> 
> Following your suggestion, made a change [2420d69](https://github.com/status-im/nimbus-eth2/commit/2420d69011879610ae012b4a078091bc7fe99c5f) to skip blob fetching when the preset is minimal. This means that all block proposals will be invalid for EIP4844-minimal, but as we've agreed elsewhere, EIP4844 is mainnet-preset only for now.

Seems fine, yeah.

Will wait to see CI a bit, but inclined to merge at this point, and can build on it in future PRs.

tersec commented on scripts/launch_local_testnet.sh:
```diff
@@ -785,6 +785,8 @@ DEPOSIT_CONTRACT_ADDRESS: ${DEPOSIT_CONTRACT_ADDRESS}
 ETH1_FOLLOW_DISTANCE: 1
 ALTAIR_FORK_EPOCH: 1
 BELLATRIX_FORK_EPOCH: 2
+CAPELLA_FORK_EPOCH: 3
```
This works, Even without separate VC support?

tersec replied:
I did this and it worked fine a while ago but only after disabling the separate VC validators and only using the built-in BN validators.

henridf replied:
Oops, this is unintentional. Juggling between branches and my own tests. Thanks for the catch.

henridf replied:
311ddd19fd0de3fd57837e5d2b23d60f6554c2de

henridf committed:
Revert unintentional script change

henridf committed:
Merge branch 'unstable' into block-proposal
